File: sieve.lua
local function sieve(limit)
    local is_composite = {}
    for i = 0, limit do
        is_composite[i] = false
    end

    local count = 0
    local p = 2
    while p <= limit do
        if not is_composite[p] then
            count = count + 1
            local multiple = p * p
            while multiple <= limit do
                is_composite[multiple] = true
                multiple = multiple + p
            end
        end
        p = p + 1
    end
    return count
end

local LIMIT = 10000000
local result = sieve(LIMIT)
print(string.format("Primes up to %d: %d", LIMIT, result))

File: bench/sieve.py
def sieve(limit):
    # bytearray, not a list of bools -- Python's own idiomatic dense-flag-array choice (one byte
    # per flag, no per-element object boxing), the fair counterpart to AER's int32 typed array.
    is_composite = bytearray(limit + 1)

    count = 0
    p = 2
    while p <= limit:
        if not is_composite[p]:
            count += 1
            multiple = p * p
            while multiple <= limit:
                is_composite[multiple] = 1
                multiple += p
        p += 1
    return count

LIMIT = 13000000
result = sieve(LIMIT)
print("Primes up to {}: {}".format(LIMIT, result))
